treq/testing.py

```diff
@@ -205,3 +207,218 @@ def __init__(self, resource):
+
+        return (set(self._headers.keys()).issubset(set(compare_to.keys())) and
+                all([set(v).issubset(set(compare_to[k]))
+                     for k, v in self._headers.items()]))
+
+    def __ne__(self, other_headers):
+        return not self.__eq__(other_headers)
+
+
+class RequestSequence(object):
+    """
+    Takes a sequence of::
+
+        [((method, url, params, headers, data), (code, headers, body)),
+         ...]
+
+    Expects the requests to arrive in sequence order.  If there are no more
+    responses, or the request's paramters do not match the next item's expected
+    request paramters, raises :obj:`AssertionError`.
+
+    For the expected request arguments::
+
+    - ``method`` should be normalized to lowercase.
+    - ``url`` should be normalized as per the transformations in
+        https://en.wikipedia.org/wiki/URL_normalization that (usually) preserve
+        semantics.  A url to `http://something-that-looks-like-a-directory`
+        would be normalized to `http://something-that-looks-like-a-directory/`
+        and a url to `http://something-that-looks-like-a-page/page.html`
+        remains unchanged.
+    - ``params`` is a dictionary mapping `str` to `lists` of `str`
+    - ``headers`` is a dictionary mapping `str` to `lists` of `str` - note that
+        :obj:`twisted.web.client.Agent` may adds its own headers though, which
+        are not guaranteed (for instance, `user-agent` or `content-length`),
+        so it's better to use some kind of matcher like :obj:`HasHeaders`.
+    - ``data`` is a `str`
+
+    For the response::
+
+    - ``code`` is an integer representing the HTTP status code to return
+    - ``headers`` is a dictionary mapping `str` to `str` or `lists` of `str`
+    - ``body`` is a `str`
+
+    :ivar list sequence: The sequence of expected request arguments mapped to
+        stubbed responses
+    :ivar async_failure_reporter: A callable that takes a single message
+        reporting failures - it's asynchronous because it cannot just raise
+        an exception - if it does, :obj:`Resource.render` will just convert
+        that into a 500 response, and there will be no other failure reporting
+        mechanism.
+    """
+    def __init__(self, sequence, async_failure_reporter):
+        self._sequence = sequence
+        self._async_reporter = async_failure_reporter
+
+    def consumed(self):
+        """
+        :return: `bool` representing whether the entire sequence has been
+            consumed.  This is useful in tests to assert that the expected
+            requests have all been made.
+        """
+        return len(self._sequence) == 0
+
+    @contextmanager
+    def consume(self, sync_failure_reporter):
+        """
+        Usage::
+
+            sequence_stubs = RequestSequence([...])
+            stub_treq = StubTreq(StringStubbingResource(sequence_stubs))
+            with sequence_stubs.consume():
```

Comment:
Oops!  This example is wrong.  `sequence_stubs.consume()` *requires* a `sync_failure_reporter` and this doesn't pass one.